packages/core/src/analytics/__tests__/dispatch.test.ts

```diff
@@ -0,0 +1,150 @@
+  event: { type: 'screen' },
+})
+describe('Dispatch', () => {
+  beforeEach(() => {
+    jest.resetAllMocks()
+    dispatchSingleSpy.mockImplementationOnce((ctx) => Promise.resolve(ctx))
+    invokeCallback.mockImplementationOnce((ctx) => Promise.resolve(ctx))
+    dispatchSpy.mockImplementationOnce((ctx) => Promise.resolve(ctx))
+    queue = new EventQueue(new PriorityQueue(4, []))
+    queue.isEmpty = jest.fn().mockReturnValue(false)
+    emitter = new Emitter()
+  })
+
+  it('should not dispatch if client is currently offline and retries are *disabled* for the main event queue', async () => {
+    isOnline.mockReturnValue(false)
+    isOffline.mockReturnValue(true)
+
+    const ctx = await dispatch({ type: 'screen' }, queue, emitter, {
+      retryQueue: false,
+    })
+    expect(ctx).toEqual(screenCtxMatcher)
+    const called = Boolean(
+      dispatchSingleSpy.mock.calls.length || dispatchSpy.mock.calls.length
+    )
+    expect(called).toBeFalsy()
+  })
+
+  it('should be allowed to dispatch if client is currently offline and retries are *enabled* for the main event queue', async () => {
+    isOnline.mockReturnValue(false)
+    isOffline.mockReturnValue(true)
+
+    await dispatch({ type: 'screen' }, queue, emitter, {
+      retryQueue: true,
+    })
+    const called = Boolean(
+      dispatchSingleSpy.mock.calls.length || dispatchSpy.mock.calls.length
+    )
+    expect(called).toBeTruthy()
+  })
+
+  it('should call dispatchSingle correctly if queue is empty', async () => {
+    queue.isEmpty = jest.fn().mockReturnValue(true)
+    await dispatch({ type: 'screen' }, queue, emitter)
+    expect(dispatchSingleSpy).toBeCalledWith(screenCtxMatcher)
+    expect(dispatchSpy).not.toBeCalled()
+  })
+
+  it('should call dispatch correctly if queue has items', async () => {
+    await dispatch({ type: 'screen' }, queue, emitter)
+    expect(dispatchSpy).toBeCalledWith(screenCtxMatcher)
+    expect(dispatchSingleSpy).not.toBeCalled()
+  })
+
+  it('should only call invokeCallback if callback is passed', async () => {
+    await dispatch({ type: 'screen' }, queue, emitter)
+    expect(invokeCallback).not.toBeCalled()
+
+    const cb = jest.fn()
+    await dispatch({ type: 'screen' }, queue, emitter, { callback: cb })
+    expect(invokeCallback).toBeCalledTimes(1)
+  })
+  it('should call invokeCallback with correct args', async () => {
+    const cb = jest.fn()
+    await dispatch({ type: 'screen' }, queue, emitter, {
+      callback: cb,
+    })
+    expect(dispatchSpy).toBeCalledWith(screenCtxMatcher)
+    expect(invokeCallback).toBeCalledTimes(1)
+    const [ctx, _cb] = invokeCallback.mock.calls[0]
+    expect(ctx).toEqual(screenCtxMatcher)
+    expect(_cb).toBe(cb)
+  })
+
+  // TODO: Inconsistent behavior? This seems like a bug.
+  it('should have inconsistent timeout behavior where the delay is different based on whether timeout is explicitly set to 1000 or not', async () => {
```

Comment:
cc @pooyaj

Reply:
I think this is intentional, 300ms default, but it is configurable.